op-chain-ops/interopgen/deploy.go

```diff
@@ -0,0 +1,322 @@
+			L2CoreDeployConfig: genesis.L2CoreDeployConfig{
+				L1ChainID: cfg.ChainID.Uint64(),
+			},
+		},
+		DevL1DeployConfig: cfg.DevL1DeployConfig,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to build L1 genesis template: %w", err)
+	}
+	allocs, err := l1Host.StateDump()
+	if err != nil {
+		return nil, fmt.Errorf("failed to dump L1 state: %w", err)
+	}
+
+	if err := noDeployed(allocs, sysGenesisDeployer); err != nil {
+		return nil, fmt.Errorf("unexpected deployed account content by L1 genesis deployer: %w", err)
+	}
+
+	for addr, amount := range cfg.Prefund {
+		acc := allocs.Accounts[addr]
+		acc.Balance = amount
+		allocs.Accounts[addr] = acc
+	}
+
+	l1Genesis.Alloc = allocs.Accounts
+
+	// Insert an empty beaconchain deposit contract with valid empty-tree prestate.
+	// This is part of dev-genesis, but not part of scripts yet.
+	beaconDepositAddr := common.HexToAddress("0x1111111111111111111111111111111111111111")
+	if err := beacondeposit.InsertEmptyBeaconDepositContract(l1Genesis, beaconDepositAddr); err != nil {
+		return nil, fmt.Errorf("failed to insert beacon deposit contract into L1 dev genesis: %w", err)
+	}
+
+	return &L1Output{
+		Genesis: l1Genesis,
+	}, nil
+}
+
+func completeL2(l2Host *script.Host, cfg *L2Config, l1Block *types.Block, deployment *L2Deployment) (*L2Output, error) {
+	deployCfg := &genesis.DeployConfig{
+		L2InitializationConfig: cfg.L2InitializationConfig,
+		L1DependenciesConfig: genesis.L1DependenciesConfig{
+			L1StandardBridgeProxy:       deployment.L1StandardBridgeProxy,
+			L1CrossDomainMessengerProxy: deployment.L1CrossDomainMessengerProxy,
+			L1ERC721BridgeProxy:         deployment.L1ERC721BridgeProxy,
+			SystemConfigProxy:           deployment.SystemConfigProxy,
+			OptimismPortalProxy:         deployment.OptimismPortalProxy,
+			DAChallengeProxy:            common.Address{}, // unsupported for now
+		},
+	}
+	// l1Block is used to determine genesis time.
+	l2Genesis, err := genesis.NewL2Genesis(deployCfg, l1Block)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build L2 genesis config: %w", err)
+	}
+
+	allocs, err := l2Host.StateDump()
+	if err != nil {
+		return nil, fmt.Errorf("failed to dump L1 state: %w", err)
+	}
+
+	if err := noDeployed(allocs, sysGenesisDeployer); err != nil {
+		return nil, fmt.Errorf("unexpected deployed account content by L2 genesis deployer: %w", err)
+	}
+
+	for addr, amount := range cfg.Prefund {
+		acc := allocs.Accounts[addr]
+		acc.Balance = amount
+		allocs.Accounts[addr] = acc
+	}
+
+	l2Genesis.Alloc = allocs.Accounts
+	l2GenesisBlock := l2Genesis.ToBlock()
+
+	rollupCfg, err := deployCfg.RollupConfig(l1Block, l2GenesisBlock.Hash(), l2GenesisBlock.NumberU64())
+	if err != nil {
+		return nil, fmt.Errorf("failed to build L2 rollup config: %w", err)
+	}
+	return &L2Output{
+		Genesis:   l2Genesis,
+		RollupCfg: rollupCfg,
+	}, nil
+}
+
+func noDeployed(allocs *foundry.ForgeAllocs, deployer common.Address) error {
```

Comment:
trying to think of a better name for this, `confirmNoUnexpectedDeployments`? Maybe that's too Java...

Reply:
Changed it to `ensureNoDeployed` and added a doc comment with further info.